### DISCUSSION TITLE
Python documentation questions / request for improvements

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#10193] [General]
I think the Python API docs & reference could greatly benefit from more detailed explanations of the differences between the various ways of running SQL statements.

For example, the reference contains 
```
duckdb.execute(query: str, parameters: object = None, multiple_parameter_sets: bool = False, connection: duckdb.DuckDBPyConnection = None) → duckdb.DuckDBPyConnection
Execute the given SQL query, optionally using prepared statements with parameters set

duckdb.from_query(query: str, alias: str = '', connection: duckdb.DuckDBPyConnection= None) → duckdb.duckdb.DuckDBPyRelation
Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, otherwise run the query as-is.

duckdb.query(query: str, alias: str = '', connection: duckdb.DuckDBPyConnection = None) → duckdb.duckdb.DuckDBPyRelation
Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, otherwise run the query as-is.

duckdb.query_df(df: pandas.DataFrame, virtual_table_name: str, sql_query: str, connection: duckdb.DuckDBPyConnection = None) → duckdb.duckdb.DuckDBPyRelation
Run the given SQL query in sql_query on the view named virtual_table_name that contains the content of DataFrame df

duckdb.sql(query: str, alias: str = '', connection: duckdb.DuckDBPyConnection = None) → duckdb.duckdb.DuckDBPyRelation
Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, otherwise run the query as-is.
```

```
class duckdb.DuckDBPyConnection

execute(self: duckdb.duckdb.DuckDBPyConnection, query: str, parameters: object = None, multiple_parameter_sets: bool = False) → duckdb.duckdb.DuckDBPyConnection
Execute the given SQL query, optionally using prepared statements with parameters set

from_query(self: duckdb.duckdb.DuckDBPyConnection, query: str, *, alias: str = '', params: object = None) → duckdb.duckdb.DuckDBPyRelation
Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, otherwise run the query as-is.

query(self: duckdb.duckdb.DuckDBPyConnection, query: str, *, alias: str = '', params: object = None) → duckdb.duckdb.DuckDBPyRelation
Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, otherwise run the query as-is.

sql(self: duckdb.duckdb.DuckDBPyConnection, query: str, *, alias: str = '', params: object = None) → duckdb.duckdb.DuckDBPyRelation
Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, otherwise run the query as-is.
```

```
class duckdb.DuckDBPyRelation

execute(self: duckdb.duckdb.DuckDBPyRelation) → duckdb.duckdb.DuckDBPyRelation
Transform the relation into a result set

query(self: duckdb.duckdb.DuckDBPyRelation, virtual_table_name: str, sql_query: str) → duckdb.duckdb.DuckDBPyRelation
Run the given SQL query in sql_query on the view named virtual_table_name that refers to the relation object
```

which leaves me with many questions:
(1) is `duckdb.from_query == duckdb.query == duckdb.sql` ? 
(2) what do the `alias`, `parameters` and `multiple_parameter_sets` parameters do? (I figured out what `parameters` does, but it still couldn't harm to add it to the description / narrow down the parameter type)
(3) Why does `duckdb.query` not allow prepared statements, but `connection.query` does? (And takes an extra `params` argument for it (why is it `params` and not `parameters` like in `execute`?))
(4) Why do the `.execute`s return `DuckDBPyConnection`s? Which connection? If it's about laziness vs eagerness, why isn't a results object of sorts returned instead?
(5) Is `duckdb.X(*, 
 **, connection=conn) == conn.X(*, **)`?
(6) What does `transform the relation into a result set` do if the input and output are of the same type? Since the function name attached to this description is `execute`, wouldn't it make more sense if a `DuckDBPyConnection` was returned like for all other `execute` functions?
(7) How is `self` used in `DuckDBPyRelation.query`? What is `virtual_table_name: str` and why does it play such a central role to come before the query string? What does it mean that "virtual_table_name refers to the relation object"? 
(8) What can `query_df` do that the other functions can't? Why does this function only exist as static functions unlike the other functions quoted here?
((9) Why do the `DuckDBPyConnection` functions include a ` *` in their signature but the static functions don't? -- I guess that's just legacy)

I'm not playing dumb here, I honestly don't know the answer to these questions (except maybe (1) which I'm pretty sure has the answer yes) despite reading all docs under the Python header on the duckdb website. 

I appreciate all the great work that's being done on duckdb and understand that writing docs is not important as adding and improving functionality. Nevertheless, I am sure even more people would start loving duckdb if the entry barriers were lowered by clarifying some of the above.

**Comment by Tishj** [comment]
Thanks, I agree that these questions are not really answered by the documentation.
Your questions will hopefully allow us to improve the docs to answer these questions 👍 

1. Yes, they are just aliases for the same functionality:
```c++
	DefineMethod({"sql", "query", "from_query"}, m, &PyConnectionWrapper::RunQuery,
	             "Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, "
	             "otherwise run the query as-is.",
	             py::arg("query"), py::arg("alias") = "", py::arg("connection") = py::none());
```

2. `alias` sets the name on the produced relation, which is necessary for operations like a join
`parameters` and `params` serve the same purpose. This is where you pass in the iterable that provides the values for the prepared parameters.
For example:
```py
import duckdb

res = duckdb.execute("select 42, ?", ['hello']).fetchall()
print(res)
# [(42, 'hello')]
```
I'm not entirely sure why `multiple_parameter_sets` is exposed to be honest, if true it acts like `executemany`
Which will execute the query multiple times, using the set of parameters:
```py
import duckdb

duckdb.execute("create table foo (a int, b varchar)")
duckdb.execute("insert into foo select 42, ?", [['hello'], ['test']], multiple_parameter_sets=True)
res = duckdb.table('foo').fetchall()
print(res)
# [(42, 'hello'), (42, 'test')]
```
So `duckdb.execute(..., multiple_parameter_sets=True)` is equivalent to `duckdb.executemany(...)`

3. Sort of an oversight, sort of a work in progress, been needing to get around to rectify that.

4. Almost every method on `DuckDBPyConnection` returns itself, to enable chaining of methods. `execute` is no exception.
This is also not really by choice, `execute` is part of the DBAPI (https://peps.python.org/pep-0249/)
Instead the created result set is part of the DuckDBPyConnection and can be interacted with through the various fetch methods, as well as `.df` `.pl`, `.arrow` etc..

5. Yes, the `connection=con` argument allows you to use a specified connection, by default it uses what we call the `default_connection` which is just an in-memory connection created at startup

6. Similar to the `DuckDBPyConnection`, the result set is stored is stored inside the relation. Relations are lazy so execution does not happen until requested, using `.execute` on the relation acts like that request.
This happens under the hood as well when you fetch as `.df()` or `.fetchall()` for example, though the path is different.
Using `.execute` will materialize the result in an intermediate representation, whereas using a "fetch" method lets us know directly what the resulting shape should be, so we can often skip this intermediate representation.

Thats one of the benefits of using the Relation objects over using `duckdb.execute`

7. `self` is used because you're creating a new relation on top of the existing relation, think for example about a projection:
```py
rel = duckdb.query("select i as a, 42 as b from range(10) t(i)")
rel2 = rel.query("my_rel", "select b from my_rel")
res = rel2.fetchall()
print(res)
# [(42,), (42,), (42,), (42,), (42,), (42,), (42,), (42,), (42,), (42,)]
```
It's a dumb example, but it shows what it does at least ^

8. `query_df` is pretty similar to `DuckDBPyRelation.query`, the difference is that the first relation is created implicitly
```py
import pandas
df = pandas.DataFrame({'a': [1, 2, 3, 4]})
rel = duckdb.query("select * from df")
rel2 = duckdb.query("select a from rel where a%2 = 0")
res = rel2.fetchall()
print(res)
# [(2,), (4,)]
```
Is equivalent to:
```py
import pandas
df = pandas.DataFrame({'a': [1, 2, 3, 4]})
rel = duckdb.query_df(df, "my_df", "select a from my_df where a%2 = 0")
res = rel.fetchall()
print(res)
# [(2,), (4,)]
```

9. I believe that signifies that they take keyword arguments, for example:
```c++
	DefineMethod({"sql", "query", "from_query"}, m, &DuckDBPyConnection::RunQuery,
	             "Run a SQL query. If it is a SELECT statement, create a relation object from the given SQL query, "
	             "otherwise run the query as-is.",
	             py::arg("query"), py::kw_only(), py::arg("alias") = "", py::arg("params") = py::none());
```

**Comment by Tishj** [comment]
I think in the future these are better raised over at `duckdb-web`, as that deals with everything documentation related
But I've already alerted the primary maintainer of that repo to see if some of this information should be made available on the doc pages :)

**Comment by soerenwolfers** [comment]
Thanks for your answers! I'll post future documentation requests and issues over there. For now though, if I may ask some follow up questions:

2. Would it be fair to say that you don't need the `alias` argument over just assigning the intermediate queries to python variables and using those variable names in future queries, unless you want to fluent-style chain queries via multiple chained function calls  (and you also don't just want to put multiple SQL statements in a single query call)? Maybe this is best answered by a meaningful toy example and maybe that is too much to ask for in this setting. 
4&6. Is there any good reason one should ever use  `execute` over `query` (plus a final `to_df` when returning to Python, and possibly some intermediate fetches during debugging)?
7. Related to 2., wouldn't it be nicer to just do `duckdb.query("select b from rel")` after the first query? Or, if you really wanted to avoid python variables, is the `virtual_table_name` not still a duplication of what can be done via the `alias` parameter in the first query?
8. Again, why not just `query("select * from df")`? Maybe  #10191 would obviate for `query_df` entirely?


I'm aware that the Python API, and duckdb itself, are the result of evolution over quite some years, and there may not always be only one way to do something, but I'd appreciate if you or someone else could still bring some clarity to new users like me as to what you consider modern / elegant ways of using the interface, and what you consider legacy / only available to satisfy things like the DB PEP.

**Comment by lucacerone** [comment]
Asking for clarifications, to me it's not clear what's the difference between `to_df` and `df` when converting a relation to a dataframe.

**Reply by soerenwolfers** [reply]
They're aliases

**Reply by lucacerone** [reply]
Thanks!
